libraries/ms-common/src/x86/uart.c

```diff
@@ -1,13 +1,133 @@
 #include "uart.h"
+#include "log.h"
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <netinet/in.h>
+#include <netdb.h>
+#include <arpa/inet.h>
+#include <pthread.h>
+
+typedef struct {
+  int sockfd;
+  int listenfd;
+  int connectfd;
+  struct sockaddr_in addr;
+  UARTStorage storage;
+} SocketData;
+
+static SocketData sock = {
+  .sockfd = 0,
+  .listenfd = 0,
+  .connectfd = 0,
+};
+
+static void *accept_handler(void *unused);
+static void *thread_handler(void *unused);
+
+void *accept_handler(void *unused) {
+
+  pthread_t receiver_thread;
+
+  sock.connectfd = accept(sock.listenfd, (struct sockaddr*)NULL, NULL);
+  close(sock.listenfd);
+
+  sock.sockfd = sock.connectfd;
+
+  if(pthread_create(&receiver_thread, NULL, thread_handler, NULL) < 0) {
+    LOG_DEBUG("Error creating receiving thread.\n");
+  }
+
+  pthread_exit(NULL);
+}
+
+void *thread_handler(void *unused) {
+
+  int n = 0;
+  uint8_t buf[1024];
+  int done = 0;
+
+  while (!done) {
+
+    n = read(sock.sockfd, buf, sizeof(buf)-1);
+
+    if (n) {
+      buf[n] = 0;
+      sock.storage.rx_handler(buf, sizeof(buf), sock.storage.context);
+    }
+
+    if(buf[0] == 'z') {
+      done = 1;
+    }
+  }
+
+  return NULL;
+}
 
 StatusCode uart_init(UARTPort uart, UARTSettings *settings, UARTStorage *storage) {
-  return status_code(STATUS_CODE_UNIMPLEMENTED);
+
+  int one = 1;
+  pthread_t receiver_thread;
+  sock.storage.rx_handler = settings->rx_handler;
+  sock.storage.context = settings->context;
+
+  memset(&sock.addr, '0', sizeof(sock.addr));
+
+  if((sock.sockfd = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
+    LOG_DEBUG("Error: Could not create sock.\n");
+    return status_code(STATUS_CODE_INTERNAL_ERROR);
+  }
+
+  sock.addr.sin_family = AF_INET;
+  sock.addr.sin_addr.s_addr = htonl(INADDR_ANY);
+  sock.addr.sin_port = htons(5000);
```

Comment:
Doesn't this end up connecting all possible instances of UART to the same server? You can't have two independent UARTs communicating with separate programs.

For example, we have 4 boards A, B, C, D where A<->B and C<->D are communicating over UART.